api/hasura/cron/dailyUpdate.ts

```diff
@@ -0,0 +1,203 @@
+      ],
+    });
+
+    for (const epoch of updateResult.epochs) {
+      const {
+        start_date,
+        end_date,
+        sendersCount,
+        sendersToday: dailySenders,
+        allocationTotals,
+        circle,
+      } = epoch;
+      assert(circle, 'epoch somehow missing circle');
+      const epochStartDate = DateTime.fromISO(start_date);
+      const epochEndDate = DateTime.fromISO(end_date);
+      const epochTimeRemaining = epochEndDate
+        .diffNow()
+        .shiftTo('weeks', 'days', 'hours')
+        .toObject();
+
+      const sendersToday =
+        dailySenders.epoch_pending_token_gifts_aggregate.nodes.map(
+          node => node.sender.name
+        );
+      const totalAllocations =
+        allocationTotals.epoch_pending_token_gifts_aggregate.aggregate
+          ?.totalAllocations.count;
+      const tokensSent =
+        allocationTotals.epoch_pending_token_gifts_aggregate.aggregate?.sum
+          ?.sumGive.tokens;
+      const optOuts = circle?.optOuts.users_aggregate.aggregate?.count;
+      const usersAllocated =
+        sendersCount.epoch_pending_token_gifts_aggregate.aggregate?.count;
+      const optedInUsers =
+        circle.receiversTotal.users_aggregate.aggregate?.count;
+
+      const message = dedent`
+        ${circle.organization?.name} / ${circle.name}
+
+        ${epochStartDate.toISODate()} to ${epochEndDate.toISODate()}
+        Total Allocations: ${totalAllocations ?? 0}
+        ${circle.token_name || 'GIVE'} sent: ${tokensSent ?? 0}
+        Opt outs: ${optOuts ?? 0}
+        Users Allocated: ${usersAllocated ?? 0} / ${optedInUsers ?? 0}
+        epoch ending ${timeStringFromDuration(epochTimeRemaining)} from now!
+        Users that made new allocations today:
+          ${sendersToday.join(', ')}
+      `;
+
+      if (circle.telegram_id) {
+        try {
+          await sendSocialMessage({
+            message,
+            circleId: circle.id,
+            channels: { telegram: true },
+            sanitize: false,
+          });
+        } catch (e: unknown) {
+          if (e instanceof Error)
+            console.error(
+              `Telegram Daily Update error for circle #${circle.id}: ` +
+                e.message
+            );
+        }
+      }
+
+      if (circle.discord_webhook) {
+        try {
+          await sendSocialMessage({
+            message,
+            circleId: circle.id,
+            channels: { discord: true },
+            sanitize: false,
+          });
+        } catch (e: unknown) {
+          if (e instanceof Error)
+            console.error(
+              `Discord Daily Update error for circle #${circle.id}: ` +
+                e.message
+            );
+        }
+      }
+    }
+
+    res.status(200).json({ message: 'updates sent' });
+  } catch (e) {
+    res.status(401).json({
+      error: '401',
+      message: (e as Error).message || 'Unexpected error',
+    });
+  }
+}
+
+function timeExists(t: number | undefined, unit: string) {
```

Comment:
this is clever

Reply:
there are libraries for this too, like moment, and looks like luxon even has a `toRelative` func that might be helpeful